### PR TITLE
feature/se-2-type-bind-upstream

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  4+3+1.5  |    0 |    0 |      0 |         0 |
| Review             |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -   |   -     |        0 |
## Problem description
This feature has been developed by DSpace and will be in the release: DSpace7.3. The code is merged to our repository.

## Problems
The type-bind code uses methods that wasn't in our repository but was developed in the DSpace7.3 milestone.
**Solution:** I copied these methods and everything works.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/SubmissionFormConverter.java [hunk 1]
```diff
@@ -74,10 +71,6 @@ private List<SubmissionFormRowRest> getPage(DCInput[][] page, String formName) {
             rowRest.setFields(fields);
             rows.add(rowRest);
             for (DCInput dcinput : row) {
-                //skip if acl says no
-                if (!isInputAuthorized(ContextUtil.obtainCurrentRequestContext() ,dcinput)) {
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/converter/SubmissionFormConverter.java:78]: why did you remove it?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/DescribeStep.java [hunk 2]
```diff
@@ -64,9 +70,19 @@ public DataDescribe getData(SubmissionService submissionService, InProgressSubmi
 
     private void readField(InProgressSubmission obj, SubmissionStepConfig config, DataDescribe data,
                            DCInputSet inputConfig) throws DCInputsReaderException {
+        String documentTypeValue = "";
+        List<MetadataValue> documentType = itemService.getMetadataByMetadataString(obj.getItem(),
+                configurationService.getProperty("submit.type-bind.field", "dc.type"));
+        if (documentType.size() > 0) {
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/DescribeStep.java:76]: empty() !?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java [hunk 3]
```diff
@@ -70,12 +80,21 @@ public List<ErrorRest> validate(SubmissionService submissionService, InProgressS
...
-                        validateMetadataValues(mdv, input, config, isAuthorityControlled, fieldKey);
-                        if (mdv.size() > 0 && input.isVisible(DCInput.SUBMISSION_SCOPE)) {
-                            foundResult = true;
+                        // If the input is not allowed for this type, strip it from item metadata.
+                        if (!input.isAllowedFor(documentTypeValue)) {
+                            itemService.removeMetadataValues(ContextUtil.obtainCurrentRequestContext(),
+                                    obj.getItem(), mdv);
+                        } else {
+                            validateMetadataValues(mdv, input, config, isAuthorityControlled, fieldKey);
+                            if (mdv.size() > 0 && input.isVisible(DCInput.SUBMISSION_SCOPE)) {
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:89]: isEmpty?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/WorkspaceItemMatcher.java [hunk 4]
```diff
@@ -82,6 +82,31 @@ public static Matcher matchItemWithTitleAndDateIssuedAndSubject(WorkspaceItem wi
...
+     * @param series the series value eg. 11-23
+     * @return  Matcher result
+     */
+    public static Matcher matchItemWithTypeAndSeries(WorkspaceItem witem, String type, String series) {
+        return allOf(
+                // Check workspaceitem properties
+                matchProperties(witem),
+                // Check type appears or is null
+                type != null ?
+                        hasJsonPath("$.sections.traditionalpageone['dc.type'][0].value", is(type)) :
```
Comment (vidiecan) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/WorkspaceItemMatcher.java:99]: use static final to avoid duplication

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/WorkspaceItemMatcher.java (hunk 4 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/matcher/WorkspaceItemMatcher.java:99]: @vidiecan This is not my code, I copied it from the upstream

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java (hunk 3 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:89]: @vidiecan  This is not my code, I copied it from the upstream

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/DescribeStep.java (hunk 2 above)
Comment (milanmajchrak) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/DescribeStep.java:76]: @vidiecan  This is not my code, I copied it from the upstream

**Review by vidiecan — APPROVED** [review]
